### ISSUE TITLE
DNARecords integration?

### ISSUE THREAD

**Issue by iris-garden** [iris-garden/test-process#948]
<h2>amanas said:</h2>
<p>Hi hail-team !</p>
<p><strong>DNARecords</strong> has recently been released. It allows exporting a <strong>Hail</strong> expression into a dataset of tfrecords or parquet files ready to use with machine learning frameworks like Tensorflow or Pytorch.</p>
<p>The whole thing is implemented on top of <strong>Hail</strong>, so it is ready to scale as much as <strong>Hail</strong> does. Of course, it takes advantage of sparsity. And the final dataset can be generated variant wise, sample wise or both at once.</p>
<p>For more details, you can review:</p>
<ul>
<li>the docs: <a href="https://dnarecords.readthedocs.io/en/latest/" class="inline-onebox" rel="noopener nofollow ugc">DNARecords — dnarecords documentation</a>
</li>
<li>the paper: <a href="https://www.biorxiv.org/content/10.1101/2022.08.13.503863v1" class="inline-onebox" rel="noopener nofollow ugc">DNARecords: An extensible sparse format for petabyte scale genomics analysis | bioRxiv</a>
</li>
<li>the code: <a href="https://github.com/amanas/dnarecords" class="inline-onebox" rel="noopener nofollow ugc">GitHub - amanas/dnarecords: Genomics data ML ready</a>
</li>
</ul>
<p>I wonder whether it could be useful to integrate it into Hail rather than having a different package for this kind of stuff.</p>
<p>Please, let me know what you think. Hopefully, it could be quite straight forward to make this integration into <strong>Hail</strong>.</p>
<p>Regards,</p>
<p>Andrés.</p>

<h2>danking said:</h2>
<p>Hey amanas !</p>
<p>This is very cool! It’s great to see so much innovation in the storage of large sequencing datasets.</p>
<p>Unfortunately, we don’t have bandwidth to integrate this ourselves. We do welcome pull requests! There’s a few issues we’d need to address to get to a pull request:</p>
<ol>
<li>We don’t want required dependencies on new libraries, like tfrecords. I think that part is best kept as a separate Python package that users can install if they’re interested.</li>
<li>Hail’s import / export procedures are all functions with names like <code>import_X</code> and <code>export_X</code>. You’ll need to restructure your code into, for example, an <code>export_dnarecords_parquet</code> function.</li>
<li>We are moving away from Apache Spark and prefer all new functionality to include a non-Spark option. In this case, your data representation should be expressible in terms of Hail Tables (just like how you currently express it in terms of Spark Datasets). We also recently released <code>Table.write_many</code> which can be used to write many Tables in parallel. You should use this instead of a ThreadPool.</li>
<li>We can only accept a pull request that has short & fast tests that don’t depend on external libraries (e.g. tfrecords).</li>
</ol>
<p>If you have bandwidth to do the above, we can collaborate. If not, I think a separate Python package is actually a pretty good option!</p>
